packages/peregrine/src/Toasts/useToasts.js

```diff
@@ -0,0 +1,123 @@
+import { useCallback, useMemo } from 'react';
+import { useToastState, useToastDispatch } from './useToastContext';
+
+// If a toast _is not_ dismissable remove it in this many milliseconds.
+const DEFAULT_TIMEOUT = 5000;
+
+/**
+ * Generates an identifier for a toast in a pure manner. Should always return
+ * the same identifier for a given set of props.
+ */
+export const getToastId = props => {
```

Comment:
Unfortunately I wanted to base it on all the possible configurable props. Before I added the hashing function I was simply concatenating the string values (see `combined` below) and using that as the ID but it made debugging a bit difficult.

All that to say that I would prefer to keep this as is but I also understand the issue with the speed of the function given that `combined.length` is pretty much always going to depend on how big the icon function is.

Reply:
The reason this code exists is to prevent confusing the shopper with identical _looking_ toasts, right? Right now, you're not doing referential comparison against the handler functions (and it wouldn't be safe to do so anyway, since they're often bound functions) so you're only comparing on visible properties. And that makes sense to me. Why not make this function just deduplicate _visually_ identical toasts?

Reply:
Oh I see why we are confused :D This does indeed look at all the props passed to `addToast` and therefor at the handlers which it _should not_. However, one of those props is the `icon` which is actually a function component returned from `react-feather`. So I could use your suggestion but I would need to add `icon` to the joined values.

I get it, and I will do it. Thanks!